05/ContactForm.js

```diff
@@ -13,96 +14,67 @@ const initialState = {
-        </label>
-    </p>
-    const RowPhone = <p>
-        <label htmlFor='phone'>Phone number:
-            <input type='tel' 
-                id='phone'
-                value={ state.phone }
-                onChange={ e => dispatch({ type: 'set-phone', payload: e.target.value }) }
-                placeholder='xxx-xxx-xxx' 
-                pattern='[0-9]{3}-[0-9]{3}-[0-9]{3}' />
-        </label>
-    </p>
-    const RowSubject = <p>
-        <label htmlFor='subject'>Subject*:
-            <input type='text' 
-                id='subject'
-                value={ state.subject }
-                onChange={ e => dispatch({ type: 'set-subject', payload: e.target.value }) }
-                placeholder='minimum five characters' 
-                minLength='5' 
-                required />
-        </label>
-    </p>
-    const RowMessage = <p>
-        <label htmlFor='message'>Message*:
-            <textarea 
-                id='message'
-                value={ state.message }
-                onChange={ e => 
-                    dispatch({ type: 'set-message', payload: e.target.value }) }              
-                rows="4" 
-                cols="50" 
-                required />
-        </label>
-    </p>
-    
-    return <form onSubmit={ e => {
+    const RowFirstName = {
+        labelText: 'First name: *', tag: 'input', type: 'text', id: 'firstName', 
+        value: state.firstName, onChangeHandler: dispatch, actionType: 'set-firstName', 
+        placeholder: 'minimum two characters',  minLength: '2', required: true
+    }
+    const RowLastName = {
+        labelText: 'Last Name: *', tag: 'input', type: 'text', id: 'lastName', 
+        value: state.lastName, onChangeHandler: dispatch, actionType: 'set-lastName', 
+        placeholder: 'minimum two characters',  minLength: '2', required: true
+    }
+    const RowEmail = {
+        labelText: 'Email: *', tag: 'input', type: 'email', id: 'email', value: state.email,
+        onChangeHandler: dispatch, actionType: 'set-email', 
+        placeholder: 'valid email address',  minLength: '3', required: true
+    }
+    const RowPhone = {
+        labelText: 'Phone number:', tag: 'input', type: 'tel', id: 'tel', value: state.phone,
+        onChangeHandler: dispatch, actionType: 'set-phone', pattern: '[0-9]{3}-[0-9]{3}-[0-9]{3}',
+        placeholder: 'xxx-xxx-xxx',  minLength: '', required: false
+    }
+    const RowSubject = {
+        labelText: 'Subject: *', tag: 'input', type: 'text', id: 'subject', value: state.subject,
+        onChangeHandler: dispatch, actionType: 'set-subject',
+        placeholder: 'minimum five characters',  minLength: '5', required: true
+    }
+    const RowMessage = {
+        labelText: 'Message: *', tag: 'textarea', id: 'message', value: state.message,
+        onChangeHandler: dispatch, actionType: 'set-message', rows: '4', cols: '50',
+        placeholder: 'minimum ten characters',  minLength: '10', required: true
+    }
```

Comment:
Spokojnie można to było wrzucić do tablicy i wykorzystać `.map()` do generowania pól formularze ;)